### ISSUE TITLE
API spike tags

### ISSUE THREAD

**Issue by SimonLab** [emfoundation/ce100-app#890]
I'm opening this new issue just for the tags. It's linked to #889 but I think we have already a few points discussed on the other issue and want to avoid mixing everything, bref here are my thought on how to manage tags:

Instead of keeping two sources of data for the tags (DAM API and Postgres database) that we will need to synchronise regularly, I would prefer to have a unique origin for the tags and use directly the API.
Synchronising tags is not an easy task as we will need to created a process that check regularly the state of the API. This can generate duplicate, outdated tags and conflicts, with only one source we avoid already these risks.

We will need first to update the organisation and challenges to match the tags give by the API. We will need to create a small script that we can run only one time.
Then when a new organisation or challenges is created we can directly save the ids of the tags given by the API.

The more complicated part will be when a tag is deleted form DAM. We need to agree first what "deleted" means. Do you want the ability to remove completely a tag from the DAM database ie from the API or do you want to do a soft delete and having the possibility to make a tag inactive? Having inactive tags will keep a reference of the tag which could make it easier for the application to display the right tags.
If you decide to be able to remove completely a tag from DAM we could:
- keep the reference to the id of the deleted tags in orgs and challenges in Postgres but make sure we don't filter or display the tag (we can check first if the tag still exist with the API)
- When a tag is deleted, DAM could trigger a post request to ce-100 application which will remove all the references of the id of the tag in the Postgres database. This method is a bit cleaner than the first one but needs more work as we will need an authentication method to not let anyone remove references.
- We could also provide a GET API endpoint to DAM which tell how manay references exist to a specific tag. This can help DAM user to see the effects of deleting a tag. This is again a bit more work and I'm not sure also if it's an essential feature for DAM, I let you decide.

From a UI perspective we also need to decide if we want to notify the user of ce-100 if a tag they use has been deleted. I'm not sure if this is a main feature to implement. Maybe an easy solution will be first to send an email to all your users each time or before a tag is deleted on DAM, something like this?: 

"Hello, the tag "tag1" has been removed from our database, if you want to replace or add other tags please make sure to edit your challenges and your organisation, thanks"

Let me know your thought @asimonw @georgemillard thanks

**Comment by asimonw** [comment]
@SimonLab thanks a lot for your analysis. We will get back to you after we've weighed the different options.

**Comment by SimonLab** [comment]
A small question concerning the "Location" cagegory:
![image](https://user-images.githubusercontent.com/6057298/30442562-bb732f78-9974-11e7-8ae3-4448c07fc788.png)

It doesn't seem to be on the list from the API request ```/api/tag-groups/```. Do you need to update the DAM API, or is the location group will be removed.
I just want also to confirm if the location is kept that we still want to display it at the end of the list? see https://github.com/emfoundation/ce100-app/issues/655#issue-206789638

**Comment by asimonw** [comment]
@SimonLab location is actually provided by a different field, as you can see in the doc we provided. This means that the last tag group in the select tags screen should be pulled from a different source.

But actually, we will have to review location before we can do the integration (especially the UI), because the list exposed by the DAM is more general and will be too big to be displayed as it currently is in the app. I'll create a separate issue for that as well. Thanks for reminding me 👍

**Comment by SimonLab** [comment]
Thanks for the reminder of the api. The location are now extracted from the following endpoint: ```/api/countries/```. This make me think that as we split the location from the tags we will need to add a specific field in our database to distinguish the tags' id from the locations'id. I will create an issue for this too :+1:

**Comment by georgemillard** [comment]
@SimonLab Regarding managing tags on the DAM. We can set things up so that:

FOR NOW
1. Only DAM Admins can delete tags
2. Tags have a 'deprecated' field, so they can be marked as no longer in use
3. This field will be exposed in the API so the CE100 app knows which tags are deprecated, and can handle this appropriately (ie letting Orgs know the tags are soon to be withdrawn)

For LATER (new issue)
4. A GET Api on the CE100 App would allow DAM Admins to know which tags are in use
5. A POST Api on the CE100 App would allow the DAM to update appropriately

**Comment by SimonLab** [comment]
@georgemillard sounds like a good plan to me :+1: 
So to confirm, the tags won't be deleted yet only marked as "deprecated" until the steps 4 and 5 are implemented, is this right @georgemillard @asimonw ?

**Comment by georgemillard** [comment]
Thanks @SimonLab, yes that is how we would do it, and I can set it up so that non-Admin users of the DAM can create Tags but not delete them, until we have in place a system to handle this (steps 4 and 5).

**Comment by asimonw** [comment]
@SimonLab Something to look into: it would be great if deprecated tags are still displayed and can still be browsed by (e.g. in Explore), but don't appear anymore in tag select screens in org or challenge forms. Edge case to be resolved: what happens when a user edits an org/challenge which is already using a deprecated tag? (to be discussed)

**Comment by SimonLab** [comment]
The tags are now fetched with the API, the deprecated tags will be cover on a specific issue
